### PR TITLE
[3.9] convert `JToolbarHelper` to a namespaced class name

### PR DESCRIPTION
**Pull request by clayfreeman**
### Summary of Changes

`JToolbarHelper` is now `Joomla\CMS\Toolbar\ToolbarHelper`. This change is necessary to further the adoption of namespaced naming schemes as opposed to the alternative old style.

### Testing Instructions

1. Check for any change in behavior of code that is presently using the `JSubMenuHelper` class name.
2. Check for any change in behavior of code that is presently using the `JToolbarHelper` class name.
3. Change an instance of `JToolbarHelper` to `Joomla\CMS\Toolbar\ToolbarHelper` in some arbitrary code and check for any change in behavior.

### Expected result

No change in behavior should occur as a result of the outlined testing methodology. The use of `JToolbarHelper` should continue to work like normal since it has been aliased to the namespaced edition of the class. The functionality of the namespaced edition of the class should be identical to that of the old version.

Since `JToolbarHelper` registered `JSubMenuHelper` with `JLoader`, `administrator/index.php` now requires `subtoolbar.php` so that the class can be made available. This was the preferred method of action as opposed to converting it to a namespaced variant since the class is deprecated.

### Documentation Changes Required

Yes. Any recommendation of `JToolbarHelper` should be changed in favor of using `Joomla\CMS\Toolbar\ToolbarHelper`.

### CODE REVIEW

**Comment by mbabker** [comment]
We've moved this class in 4.0 to the libraries, TBH I don't remember anymore if there's a reason we can't or shouldn't do this for 3.x.  // @wilsonge @laoneo

**Comment by mbabker** [comment]
(Also if we do this for 3.9 it should be placed in the `Joomla\CMS\Toolbar` namespace as it is in 4.0 so we don't have several instances of the class being renamed between releases)

**Comment by clayfreeman** [comment]
@mbabker I dug a little deeper and see that now; I'll revise the location and update my PR comment.

Some sort of compatibility layer should be available before 4.0 hits so that developers have time to switch to the new namespaced variant of the class name.

**Comment by mbabker** [comment]
> Some sort of compatibility layer should be available before 4.0 hits so that developers have time to switch to the new namespaced variant of the class name.

4.0 is still supporting all the global class names except those explicitly deprecated for removal in 4.0 (basically some of the early migrated Framework packages or classes that are explicitly deprecated and will be removed).  So, there is no need to migrate calls to `JToolbarHelper` to the namespaced variant for a while.

**Comment by clayfreeman** [comment]
While this is true, Joomla 4.0 is a bit far off. If it were being released tomorrow, this would be a non-issue. But, since it isn't, `JToolbarHelper` is one of the few classes that you *can't* find a namespaced variant for when writing an extension.

For developers that are trying to use the new naming scheme exclusively, `JToolbarHelper` is a bit of a blemish on that record.

**Comment by mbabker** [comment]
Oh, I agree it should be done.  Just pointing out that there is no rush in completely moving over to all namespaced classes, and that even in 4.0 the old class names will work for the most part.  Using namespaced classes keeps your IDE from groaning about missing classes (without generating a stubs file), that's probably the biggest benefit right now.

**Comment by travisrisner** [comment]
I have tested this item :white_check_mark: successfully on 5b7d763db40b4bacd2e8f3f4046659b72e7400c7.

File: administrator/index.php [hunk 1]
```diff
@@ -39,7 +39,7 @@
 
 require_once JPATH_BASE . '/includes/framework.php';
 require_once JPATH_BASE . '/includes/helper.php';
-require_once JPATH_BASE . '/includes/toolbar.php';
+require_once JPATH_BASE . '/includes/subtoolbar.php';
```
Comment (laoneo) [inline administrator/index.php:42]: Can you also namespace the JSubMenuHelper class, so we have all converted? Thanks.

Comment (mbabker) [reply administrator/index.php:42]: It's deprecated and gone in 4.0.  No point in namespacing it.

Comment (laoneo) [reply administrator/index.php:42]: Ah missed that.

**Comment by jeremyvii** [comment]
This item has been tested successfully :white_check_mark:  on  5b7d763

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21372">issues.joomla.org/tracker/joomla-cms/21372</a>.</sub>

**Review by laoneo — APPROVED** [review]

**Comment by NickSdot** [comment]
This change is causing issues in 3pd extensions we use.

```
Warning: require_once(/www/administrator/includes/toolbar.php): failed to open stream: No such file or directory in www/components/com_payplans/xiframework/base/abstract/helper/toolbar.php on line 11

Fatal error: require_once(): Failed opening required 'www/administrator/includes/toolbar.php' (include_path='.:/usr/local/lib/php') in www/components/com_payplans/xiframework/base/abstract/helper/toolbar.php on line 11
```

Is there any plan to consider this as a B/C breaking change and revert?

**Comment by clayfreeman** [comment]
Get in touch with the third party developer to have them fix it. It should be as simple as surrounding the `require` in a conditional.

Alternatively, take the line out of the file yourself.

**Comment by joomdonation** [comment]
> It should be as simple as surrounding the `require` in a conditional.

We just use the below command, works with both Joomla older version and latest version
```php
JLoader::register('JToolbarHelper', JPATH_ADMINISTRATOR . '/includes/toolbar.php');
```

**Comment by clayfreeman** [comment]
That’s quite a bit more elegant than my suggestion!

**Pull request merged**